functions/index.js

```diff
@@ -248,46 +260,93 @@ exports.onUsersChange = functions.database.ref(parentRoomID + "{roomID}/users").
+  return admin.database().ref(parentRoomID + roomID).child("users").once('value', (snapshot) => {
     snapshot.forEach((child) => {
-      updates[child.val()] = 0;
+      updates[child.val()] = value;
+    });
+  }).then(() => { 
+    return admin.database().ref(parentRoomID + roomID).child(node).once('value', function(snapshot) {
+      snapshot.ref.set(updates);
     });
   });
-
-  admin.database().ref(parentRoomID + roomID).child(node).once('value', function(snapshot) {
-    snapshot.ref.set(updates);
-  });
-
-  return;
 }
 
-function updateUser(userID, child, position) {
+function updateUser(userID) {
   return admin.database().ref("users/").once('value', (snapshot) => {
     if(snapshot.hasChild(userID)) {
-      admin.database().ref("users/" + userID + "/stars").transaction((currentValue) => {
-          return currentValue + child.val();
-      });
-      admin.database().ref("users/" + userID + "/trophies").transaction((currentValue) => {
-          return currentValue + getRankFromPosition(position);
+      let newTrophies;
+      let totalMatches;
+
+      return admin.database().ref("users/" + userID + "/trophies").transaction((currentValue) => {
+        newTrophies = currentValue + minRank;
+        return newTrophies;
+      })
+      .then(() => {
+        return admin.database().ref("users/" + userID + "/currentLeague").transaction((currentValueInner) => {
+          return getLeagueFromTrophies(newTrophies);
+        });
+      })
+      .then(() => {
+        return admin.database().ref("users/" + userID + "/totalMatches").transaction((currentValue) => {
+          totalMatches = currentValue + 1;
+          return totalMatches;
+        });
+      })
+      .then(() => {
+        return admin.database().ref("users/" + userID + "/averageRating").transaction((currentValueInner) => {
+          return parseFloat(((currentValueInner*(totalMatches-1))/totalMatches).toPrecision(3));
+        });
       });
     }
   });
 }
 
+function updateDisconnectedUsers(snapshotRanking, roomID) {
+  let dict = {};
+  return admin.database().ref(parentRoomID + roomID + "/users").once('value', (snapshot) => {
+    return snapshot.forEach((child) => {
+      dict[child.val()] = child.key;
+    });
+  }).then(() => {
+      return snapshotRanking.forEach((child) => {
+        if(child.val() === -1) {
+          updateUser(dict[child.key]);
+        }
+    });
+  });
+}
+
+function checkOnlineNode(roomID) {
+  return admin.database().ref(parentRoomID + roomID + "/onlineStatus").once('value', (snapshot) => {
+    snapshot.forEach((child) => {
+      console.log(child.val());
+      if(child.val() === 0) {
+        admin.database().ref(parentRoomID + roomID + "/ranking/" + child.key).set(-1);
+      }
+    });
+  }).then(() => {return createNode(roomID, "onlineStatus", 0)});
+}
+
 exports.onStateUpdate = functions.database.ref(parentRoomID + "{roomID}/state").onWrite((change, context) => {
```

Comment:
This code looks like a mess. Consider extract method here :)

Reply:
Javascript is a mess in general ;)